### ISSUE TITLE
[FEATURE] Message Hook

### ISSUE THREAD

**Issue by tor-web3** [python-telegram-bot/python-telegram-bot#3150]
### What kind of feature are you missing? Where do you notice a shortcoming of PTB?

- I want to create a global callback for the message without affecting the normal processing of other business.
- I want to get all the conversations the bot has with the user to better understand the user's needs
- I want to record the history of all sent messages to ensure the bot has a better contextual understanding
- I want to make the bot's replies smarter by recording the message history

### Describe the solution you'd like

Whether there is a `Message Hook`, including but not limited to a conversation hook for private chat.

This hook will intercept including inline buttons, private chat messages, etc. 

This includes messages sent by users to bots, as well as messages sent by bots to users.

If this hook does not exist, we can discuss the value of the existence of this hook, the scenarios in which it is used.

### Describe alternatives you've considered

`filters.ALL` cannot intercept all messages, I have tried, such as the registered command `/lang`
```
def setup_application(dp):
    # language
    # for handler in language_handlers.handlers():
    #     dp.add_handler(handler)
    
    from tgbot.handlers.base import message_hook
    dp.add_handler(MessageHandler(filters.ALL, message_hook))
```

### Additional context

PTB: v20

**Comment by tor-web3** [comment]
There are other ways to achieve this, but it clutters the code. I think this feature should be integrated into `PTB` to be part of it.

**Comment by Bibo-Joshi** [comment]
Hi. Thanks for reaching out. For clarification: When you say "conversation" do you mean `ConversationHandler`? or what exactly do you mean by "conversation hook"?

> `filters.ALL` cannot intercept all messages, I have tried, such as the registered command `/lang`

`MessageHandler(filters.ALL, …)` handles *any* incoming message (also edited messages and (edited) channel posts). If it doesn't, then you probably have a different handler set up that already handles the message. If you can't find the issue, please providen an [MWE](../wiki/MWE).

> I want to create a global callback for the message without affecting the normal processing of other business.

To do something with all incoming updates, you can add a `TypeHandler(type=telegram.Update, callback)` to a low group. Please also see docs of [`add_handler(…, group=…)`](https://docs.python-telegram-bot.org/en/v20.0a2/telegram.ext.application.html#telegram.ext.Application.add_handler.params.group) and [this wiki page](https://github.com/python-telegram-bot/python-telegram-bot/wiki/Frequently-requested-design-patterns#how-to-handle-updates-in-several-handlers).

> as well as messages sent by bots to users.

Storing all these messages in an automated way is not completely straighgtforward, but you can get close with little effort.

1. You can subclass `HTTPXRequest`  and override the [`post`](https://docs.python-telegram-bot.org/en/v20.0a2/telegram.request.baserequest.html#telegram.request.BaseRequest.post) method to something like `result = await super().post(…); do_something_with(result); return result`. Here `result` would be json data instead of e.g. an `telegram.Message` object. However, it covers all api methods easily
2. If it's important to do you have the result as `telegram.Message`, you can subclass `telegram.Bot` and override all bot methods to do something like `result = await super().send_message(…); do_something_with(result); return result`

**Comment by tor-web3** [comment]
> To do something with all incoming updates, you can add a `TypeHandler(type=telegram.Update, callback)` to a low group. Please also see docs of [`add_handler(…, group=…)`](https://docs.python-telegram-bot.org/en/v20.0a2/telegram.ext.application.html#telegram.ext.Application.add_handler.params.group) and [this wiki page](https://github.com/python-telegram-bot/python-telegram-bot/wiki/Frequently-requested-design-patterns#how-to-handle-updates-in-several-handlers).

I looked at this wiki and `TypeHandler` is the solution I need, but it still behaves differently than I understand.

```python
from telegram import Update
from telegram.ext import ContextTypes, ApplicationHandlerStop, TypeHandler, Application


async def language_command(update: Update, context: ContextTypes.DEFAULT_TYPE):
    u, created = await get_user_and_created(update, context)
    await update.message.reply_text(text="change_language")

async def callback(update: Update, context: ContextTypes.DEFAULT_TYPE):
    """Handle the update"""
    # await do_something_with_this_update(update, context)
    # raise ApplicationHandlerStop # Only if you DON'T want other handlers to handle this update
    pass


app = Application.builder().token("TOKEN").build()
handler = TypeHandler(Update, callback) # Making a handler for the type Update
app.add_handler(handler) # Default is 0, so we are giving it a number below 0
# Add other handlers and start my bot.
# eg: /lang
language_command_handler = CommandHandler("lang", language_command)
app.add_handler(language_command_handler)
```

Will be processed in the order of `add_handler`, `callback` takes precedence. and ignore the execution of other `handler` programs

**Comment by tor-web3** [comment]
# solution
I solved this problem by adjusting the `group` parameter.
This is an implicit logic that PTB only handles one `handler` within the same `group`
```python
handler = TypeHandler(Update, callback)
app.add_handler(handler,-1)
```
Is my understanding accurate?

# introduce
```python
add_handler(self, handler: BaseHandler[Any, CCT], group: int = DEFAULT_GROUP)
```
## Register a handler.

TL;DR: Order and priority counts. 0 or 1 handlers per group will be used. End handling of
update with :class:`telegram.ext.ApplicationHandlerStop`.

A handler must be an instance of a subclass of :class:`telegram.ext.BaseHandler`. All
handlers
are organized in groups with a numeric value. The default group is 0. All groups will be
evaluated for handling an update, but only 0 or 1 handler per group will be used. If
:class:`telegram.ext.ApplicationHandlerStop` is raised from one of the handlers, no further
handlers (regardless of the group) will be called.

The priority/order of handlers is determined as follows:

  * Priority of the group (lower group number == higher priority)
  * The first handler in a group which can handle an update (see
    :attr:`telegram.ext.BaseHandler.check_update`) will be used. Other handlers from the
    group will not be used. The order in which handlers were added to the group defines the
    priority.

### Warning

Adding persistent :class:`telegram.ext.ConversationHandler` after the application has
been initialized is discouraged. This is because the persisted conversation states need
to be loaded into memory while the application is already processing updates, which
might lead to race conditions and undesired behavior. In particular, current
conversation states may be overridden by the loaded data.

### Args

handler (:class:`telegram.ext.BaseHandler`): A BaseHandler instance.

group (:obj:`int`, optional): The group identifier. Default is ``0``.

**Comment by tor-web3** [comment]
> Hi. Thanks for reaching out. For clarification: When you say "conversation" do you mean `ConversationHandler`? or what exactly do you mean by "conversation hook"?

The conversations I'm referring to here are business processes and don't involve any programming code.

The conversation here is like the process of communication between you and me.


> To do something with all incoming updates, you can add a `TypeHandler(type=telegram.Update, callback)` to a low group. Please also see docs of [`add_handler(…, group=…)`](https://docs.python-telegram-bot.org/en/v20.0a2/telegram.ext.application.html#telegram.ext.Application.add_handler.params.group) and [this wiki page](https://github.com/python-telegram-bot/python-telegram-bot/wiki/Frequently-requested-design-patterns#how-to-handle-updates-in-several-handlers).

User message interception I found a solution in [WIKI](https://github.com/python-telegram-bot/python-telegram-bot/wiki/Frequently-requested-design-patterns#boilerplate-code).

` BOT` Please drop me a note if I can get a similar friendly solution.

> as well as messages sent by bots to users.
> 
> Storing all these messages in an automated way is not completely straighgtforward, but you can get close with little effort.
> 
> 1. You can subclass `HTTPXRequest`  and override the [`post`](https://docs.python-telegram-bot.org/en/v20.0a2/telegram.request.baserequest.html#telegram.request.BaseRequest.post) method to something like `result = await super().post(…); do_something_with(result); return result`. Here `result` would be json data instead of e.g. an `telegram.Message` object. However, it covers all api methods easily
> 2. If it's important to do you have the result as `telegram.Message`, you can subclass `telegram.Bot` and override all bot methods to do something like `result = await super().send_message(…); do_something_with(result); return result`

The processing of the program is often not single, and may involve huge microservice applications or other business modules developed by team members.

In private chat mode, preserving the full context facilitates "AI" program processing.

I could subclass or intercept `send_message` to do what I need, but that would complicate the application layer code.

Like `add_handler`, possibly a friendlier design. If you expose `send_message_hook` to solve application layer message interception.

`send_message_hook` Just a pronoun, it can be any more appropriate name, eg: add_hook

**Comment by Bibo-Joshi** [comment]
> This is an implicit logic that PTB only handles one handler within the same group

It's well documented (you even copy-pasted the documentation in here) though 😉 

> Like `add_handler`, possibly a friendlier design. If you expose `send_message`_hook to solve application layer message interception.
> `send_message_hook` Just a pronoun, it can be any more appropriate name, eg: `add_hook`

My impression is that this is a quite specific use case. E.g. I'd argue that a more common use case is to store messages sent by the bot only in some specific occasions and not *everything*. This however is already very easy by saving the return value of `bot.send_message(…)`. IIRC you're the first one to suggest such a feature.

At least ATM I don't think that I'd liko to maintain such a functionality within PTB. an addition to [ptbcontrib](https://github.com/python-telegram-bot/ptbcontrib) can be considered. If the other dev-team members disagree, they are ofc very welcome to say so! 🙂

**Comment by tor-web3** [comment]
His necessity does take time to think about.
